File: 0875-koko-eating-bananas/0875-koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        left = 1
        right = max(piles)

        while left <= right:
            mid = (left + right) // 2
            hours = 0
            for pile in piles:
                hours += (pile + mid - 1) // mid
            
            if hours <= h:
                right = mid - 1
            else:
                left = mid + 1
        
        return left

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int right = piles[0];
        for (int pile: piles) {
            right = Math.max(right, pile);
        }

        int left = 1;
        while (left <= right) {
            int mid = (left + right) / 2;
            int hour = 0;
            for (int pile: piles) {
                hour += Math.ceil((double) pile / mid);
            }
            if (hour <= h) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return left;
    }
}